src/main/java/org/broadinstitute/hellbender/utils/read/ReadUtils.java

```diff
@@ -935,17 +872,12 @@ public static GATKRead resetOriginalBaseQualities(final GATKRead read) {
      * Check to ensure that the alignment makes sense based on the contents of the header.
      * @param header The SAM file header.
      * @param read The read to verify.
-     * @return true if alignment agrees with header, false otherwise.
+     * @return false if read is mapped to a nonexistent contig or past end of contig, true otherwise
      */
     public static boolean alignmentAgreesWithHeader(final SAMFileHeader header, final GATKRead read) {
         final int referenceIndex = getReferenceIndex(read, header);
-        // Read is aligned to nonexistent contig
-        if( ! read.isUnmapped() && referenceIndex == SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX ) {
-            return false;
-        }
-        final SAMSequenceRecord contigHeader = header.getSequence(referenceIndex);
-        // Read is aligned to a point after the end of the contig
-        return read.isUnmapped() || read.getStart() <= contigHeader.getSequenceLength();
+        return read.isUnmapped() || (referenceIndex != SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX
+                && read.getStart() <= header.getSequence(referenceIndex).getSequenceLength());
```

Comment:
Hey @davidbenjamin, this kind of refactoring involving flipping boolean logic everywhere is actually really risky and error-prone in my experience, particular if there aren't comprehensive unit tests for all of these methods. I have a feeling if I spent an hour looking at this branch, I might be able to find a new bug or two. Given that we're trying to stabilize for a release at the moment, and the benefits of doing this refactoring are pretty slight, I recommend we hold off on these kinds of changes for now.